src/dune/match_lang.ml

```diff
@@ -0,0 +1,39 @@
+open! Stdune
+
+module Ast = struct
+  type ('k, 'v) t =
+    { on : String_with_vars.t
+    ; clauses : ('k * 'v) list
+    }
+
+  let decode k v =
+    let open Dune_lang.Decoder in
+    enter
+      (let* loc, () = located (keyword "match") in
+       let* on = String_with_vars.decode in
+       let clause =
+         enter
+           (let* elt = k in
+            let* () = keyword "->" in
+            let+ v = v in
+            (elt, v))
+       in
+       let+ clauses = repeat clause in
+       ( match clauses with
+       | [] ->
+         User_error.raise ~loc
+           [ Pp.text "match expression must have at least one clause" ]
+       | _ :: _ -> () );
+       { on; clauses })
+end
+
+type 'a t = (Predicate_lang.t, 'a) Ast.t
+
+let decode f = Ast.decode Predicate_lang.decode f
```

Comment:
Is there a syntax for the match-all wildcard (`Predicate_lang.any`) in the predicate language? That will probably be necessary if we use this for eg ocaml flags.

Reply:
There is artificial syntax I suppose: `(not ())` or something. I definitely agree that adding explicit syntax for it be welcome.